File: Lab 5/Python/Task_1_Uniform.py
import numpy as np
from scipy.integrate import quad
import matplotlib.pyplot as plt
from sympy import *


def uniform_integral(a, b):
    PDF = lambda x: x**0 * 1 / (b - a)
    total_prob, error_prob = quad(lambda x: x**0 * 1 / (b - a), a, b)
    mean, error_m = quad(lambda x: x**1 * 1 / (b - a), a, b)
    expectation_xsq_fx, error_var = quad(lambda x: x**2 * 1 / (b - a), a, b)

    variance = expectation_xsq_fx - mean**2
    return [PDF, total_prob, mean, variance]


a = 1
b = 4
fs = 100

x_plot = np.arange(a, b, 1/fs)

PDF, total_probability, mean, variance = uniform_integral(a, b)
print('Total Probability of the PDF', total_probability,
      '\nMean of the PDF', mean,
      '\nVariance of the PDF', variance)

plt.figure(1)
plt.plot(x_plot, PDF(x_plot))
plt.title('Uniform PDF')
plt.xlabel('Running Variable')
plt.ylabel('Amplitude')
plt.show()

# Additional -- Finding PDF using symbolic integration
x = Symbol('x')
PDF = x**0 * 1 / (b - a)
CDF = integrate(PDF, x)
print('The CDF is', CDF)

File: Lab 5/MATLAB/Task_1_Uniform.m
% PDF of Uniform RV

b = 4;
a = 1;
fs = 10;

% Calculating probability, mean and variance using the integral formulae
% for CDF and expectation
[total_prob, mean_val,var_val] = UniformRV(b,a,fs);
disp('The total probability (should be 1):')
disp(total_prob)
disp('Mean for the distribution:')
disp(mean_val)
disp('Variance for the distribution:')
disp(var_val)

function [q_0,mean_val,var_val] = UniformRV(b,a,fs)
    x = a:1/fs:b;
    pdf = zeros(1, numel(x));
    for i = 1:1:numel(x)
        if( (x(i) >= a)&&(x(i) <= b))
            pdf(i) = 1/(b - a);
        else
            pdf(i) = 0;
        end
    end
    pdf_fn = @(t) t.^0.*1/(b - a);
    exp_x_fn = @(t)t.*(1/(b - a));
    exp_xsq_fn = @(t)t.^2.*(1/(b - a));

    figure()
    plot(x,pdf)
    axis([a-2 b+2 0 4/(b-a)])
    title("Uniform RV")
    xlabel("Running variable")
    ylabel("PDF")
    grid on

    q_0 = integral(pdf_fn,a,b);
    mean_val = integral(exp_x_fn,a,b);
    var_val = integral(exp_xsq_fn,a,b) - mean_val^2;
end